api4/data_retention.go

```diff
@@ -4,21 +4,294 @@
+		body = map[string]interface{}{
+			"policies":    policies,
+			"total_count": len(policies),
+		}
+	}
+	b, _ := json.Marshal(body)
+	w.Write(b)
+}
+
+func getPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policyId := mux.Vars(r)["policy_id"]
+	policy, err := c.App.GetRetentionPolicy(policyId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	w.Write(policy.ToJsonBytes())
+}
+
+func createPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policy, jsonErr := model.RetentionPolicyFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+		return
+	}
+	auditRec := c.MakeAuditRecord("createPolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("policy", policy)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	policy, err := c.App.CreateRetentionPolicy(policy)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	auditRec.AddMeta("policy", policy) // overwrite meta
+	auditRec.Success()
+	w.WriteHeader(http.StatusCreated)
+	w.Write(policy.ToJsonBytes())
+}
+
+func patchPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	patch, jsonErr := model.RetentionPolicyUpdateFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+	}
+	patch.Id = mux.Vars(r)["policy_id"]
+
+	auditRec := c.MakeAuditRecord("patchPolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("patch", patch)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	policy, err := c.App.PatchRetentionPolicy(patch)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	auditRec.Success()
+	w.Write(policy.ToJsonBytes())
+}
+
+func updatePolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	update, jsonErr := model.RetentionPolicyUpdateFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+	}
+	update.Id = mux.Vars(r)["policy_id"]
+
+	auditRec := c.MakeAuditRecord("updatePolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("update", update)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	policy, err := c.App.UpdateRetentionPolicy(update)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	auditRec.Success()
+	w.Write(policy.ToJsonBytes())
+}
+
+func deletePolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policyId := mux.Vars(r)["policy_id"]
+
+	auditRec := c.MakeAuditRecord("deletePolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("policy_id", policyId)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	err := c.App.DeleteRetentionPolicy(policyId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	auditRec.Success()
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func addTeamsToPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policyId := mux.Vars(r)["policy_id"]
+	var lst teamIdsList
+	jsonErr := json.NewDecoder(r.Body).Decode(&lst)
```

Comment:
In that case the client would have to upload a JSON array instead of a JSON object containing an array. I'm leaning towards keeping the request body as a JSON object so that we have the flexibility to add more keys later on if necessary. Thoughts?